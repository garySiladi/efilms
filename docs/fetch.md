# Fetch API in Economist Films
We use data fetching for getting the proper data for the website.
We can fetch:
* the whole root JSON
* a series by ID
* an episode by ID
# Fetch functions
#### getRoot()
```
getRoot().then((data) => {
    ... the whole root JSON
});
```
#### getSeriesByID(id)
```
getSeriesByID(id).then((data) => {
    ... the series with the given ID in JSON format
});
```
#### getEpisodeByID(id)
```
getEpisodeByID(id).then((data) => {
    ... the episode with the given ID in JSON format
});
```
