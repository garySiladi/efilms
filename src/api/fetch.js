// @flow
function constructFetch(type: string, id: ?number) {
  const urlString = id ? `${type}/${id}` : type;
  return fetch(`https://economist.twivel.io/api/v1/${urlString}/json`, {
    method: 'GET',
  })
  .then(response => response.json())
  .catch(err => err);
}

export function getRecommendedEpisodes(userId: string) {
  return fetch(`http://172.17.4.94/recommendations/${userId}`,
    {
      method: 'GET',
    })
  .then(response => response.json());
}

export function getRoot() {
  return constructFetch('root');
}

export function getEpisodeByID(id: number) {
  return constructFetch('episode', id);
}

export function getSeriesByID(id: number) {
  return constructFetch('series', id);
}
