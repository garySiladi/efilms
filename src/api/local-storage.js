// @flow

export type VideoProgressType = {
  episodeId: number,
  progressTime: number,
};

export const HISTORY_LIST = 'history';

export function saveVideoProgress(id: number, progress: number) {
  if (typeof (localStorage) !== 'undefined') {
    const currentHistory: ?string = localStorage.getItem(HISTORY_LIST);
    const parsedCurrentHistory = JSON.parse(currentHistory || 'null') || [];
    const indexOfVideo = parsedCurrentHistory.findIndex(element => element.episodeId === id);
    if (indexOfVideo === -1) {
      parsedCurrentHistory.push({
        episodeId: id,
        progressTime: progress,
      });
    } else {
      parsedCurrentHistory[indexOfVideo].progressTime = progress;
    }
    localStorage.setItem(HISTORY_LIST, JSON.stringify(parsedCurrentHistory));
  }
}

export function getProgressTimeById(id: number) {
  const data: ?string = localStorage.getItem(HISTORY_LIST);
  const parsedData = JSON.parse(data || 'null');
  if (parsedData) {
    const requiredEpisode = parsedData.find(episode => episode.episodeId === id);
    return requiredEpisode ? requiredEpisode.progressTime : 0;
  }
  return 0;
}

export function getLastWatchedEpisodeID(episodeIDs: Array<number>) {
  const data: ?string = localStorage.getItem(HISTORY_LIST);
  const historyData = JSON.parse(data || 'null');
  if (historyData) {
    return episodeIDs.find(episodeID =>
      historyData.every(
        histItem => histItem.episodeId !== episodeID || histItem.progressTime < 95,
      ),
    ) || episodeIDs[0];
  }
  return episodeIDs[0];
}
