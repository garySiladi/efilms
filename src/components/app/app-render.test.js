// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import App from './app';
import * as API from '../../api/fetch';

jest.mock('../side-panel/side-panel', () =>
  jest.fn(() => <div>Side Panel</div>),
);

jest.mock('../home-container/home-container', () =>
  jest.fn(() => <div>Home Container</div>),
);

const mockedLocationData = {
  query: {},
};

// $FlowFixMe
API.getRoot = jest.fn()
.mockReturnValue(
  new Promise((resolve) => {
    resolve({ shelves: [] });
  }),
);

describe('App: ', () => {
  test('renders correctly', () => {
    const tree : string = renderer.create(<App params={{}} location={mockedLocationData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
