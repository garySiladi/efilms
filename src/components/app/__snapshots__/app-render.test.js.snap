// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App:  renders correctly 1`] = `
<div
  className="app"
>
  <div>
    Home Container
  </div>
</div>
`;
